### DISCUSSION TITLE
How to install airflow in setup.py

### DISCUSSION THREAD

**Discussion by aamster** [apache/airflow#29160] [Q&A]
Airflow requires to install with constraints file such as `pip install apache-airflow --constraint <constraint`

How can this constraint file be passed to the package name in `install_requires` in disutils setup.py? I have tried many things but can't seem to get it to work. 

I should be able to do `pip install .` to install my package with airflow as a dependency but can't pass the constraint file

i tried adding

```
    cmdclass={
        'develop': PostDevelopCommand,
        'install': PostInstallCommand,
    }
```

to `setup`

with 

```
class PostDevelopCommand(develop):
    """Post-installation for development mode."""
    def run(self):
        develop.run(self)
        # PUT YOUR POST-INSTALL SCRIPT HERE or CALL A FUNCTION
        cmd = f'pip install apache-airflow==2.5.1 {get_airflow_install_args(line="apache-airflow==2.5.1")}'
        check_call(cmd.split())


class PostInstallCommand(install):
    """Post-installation for installation mode."""
    def run(self):
        install.run(self)
        # PUT YOUR POST-INSTALL SCRIPT HERE or CALL A FUNCTION
        cmd = f'pip install apache-airflow==2.5.1 {get_airflow_install_args(line="apache-airflow==2.5.1")}'
        check_call(cmd.split())
```

this runs, but the `airflow` command doesn;t work

**Comment by aamster** [comment]
install outside of setup.py instead

**Comment by potiuk** [comment]
You can't. You need to add constraints as you want in your `pip install` command (constraints can be placed for any installation - including your own package that has airflow dependencies). They should be specified as parameter of the `pip` command - they cannot be specified as 'expectations' in your package configuration.  I've been thinking about creating a PEP and proposal to packageing team of Python a proposal how similar feature can be handled without it (specifying a "golden set" of dependencies for package) but such feature does not exist. 

BTW. Airfow does not REQUIRE those constraints to be used. Constraints are the way that you can be sure that Airlfow is installed and working as repeatable installeation - preventing airlfow from errors coming from external packages released after Airflow has been released. But you are free to install airflow with newer versions of those dependencies if they were released after airflow - but it's up to you to test them that it works. This is no-different from any other python package (and the final check is on the one who installs the packages) but Airlfow adds extra protection in case you want to make sure the installation is repeatable with the versions of dependencies that were "valid" and "tested" at the moment of releasing given airflow version.

You are free to choose your path:

1) install airflow with the constraints to make sure repeatable installation works
2) install airfloow without the constraints or upgrade the dependencies if you want to update those dependencies for whatever reason (but then it's on you to make sure to test it).

It cannot be made as part of setup.py because there is no standard/PEP defined for it, so you need to add - whatever approach you take - to your scripting around it.
